### ISSUE TITLE
Consider spliting event IDs into "source" and "event" space

### ISSUE THREAD

**Issue by plafer** [0xMiden/miden-vm#1583]
As a follow-up to #1457.

Context: 
- main issue comment: https://github.com/0xPolygonMiden/miden-vm/issues/1457#issue-2471210549
- PR discussion: https://github.com/0xPolygonMiden/miden-vm/pull/1581#discussion_r1853114792

**Comment by adr1anh** [comment]
After discussing with @bobbinth, we came up with the solution of using blake3 for mapping a fully qualified event down to a single `Felt`. In particular, we derive it as `blake3(event_name) % P`. 

The naming scheme is still TBD. A simple approach for now would just be `library::event_name` though this leaves a couple of open questions
- Do we need an additional "namespace" for disambiguation?
- Should we enforce upper/lowercase naming?
- Where should the "library" name come from?

Another question is whether we want to continue supporting `u32` as IDs, or if we should transition to the named approach. This would break backwards compatibility but should slightly simplify the grammar. We could support them for the next release while marking them as deprecated to avoid too many issues with miden-base.

Finally, my comment in [comment](https://github.com/0xMiden/miden-vm/issues/2011#issuecomment-3183410712) in #2011 suggest introducing a new `emit(library::event_name)` syntax to disambiguate event names from classic strings. While this could provide a some type safety, it might complicate the grammar unnecessarily. The main reason I can think of for not going with this approach would be if we wanted to repurpose the event's string as a key for a map, for example.

**Comment by bobbinth** [comment]
I think https://github.com/0xMiden/miden-vm/pull/2075#issuecomment-3194747185 addresses most of the questions (at least implicitly) - but to make sure I answer them explicitly:

> Do we need an additional "namespace" for disambiguation?

I think it would be good to prepend the user-provided string with the namespace of the library in which the module is located. But other than that, I wouldn't do any transformations of the user-provided event name.

> Should we enforce upper/lowercase naming?

I think for event names (i.e., specified via `event` keyword) we could apply the same rules as we do for library paths. I believe there we allow using ASCII characters, underscores and `::`, and I think we require them to be lower-case.

> Where should the "library" name come from?

We should be able to get the namespace from the library currently being assembled. If it that's not easy to do at the time we compute the hash of the string, then I think it is fine to skip prepending the namespace altogether and rely just on the convention.

> Another question is whether we want to continue supporting `u32` as IDs, or if we should transition to the named approach. This would break backwards compatibility but should slightly simplify the grammar. We could support them for the next release while marking them as deprecated to avoid too many issues with miden-base.

I think it is fine to make this a breaking change for v0.18.0 release and not support the legacy format (this would be the case anyways if we go with #2068).

**Comment by adr1anh** [comment]
Thanks for the feedback, this and the PR comment answers my questions. For now, I'll come back to #2075 and focus on using the stack for emit operations.

**Comment by plafer** [comment]
> I think it would be good to prepend the user-provided string with the namespace of the library in which the module is located. But other than that, I wouldn't do any transformations of the user-provided event name.

While cleaner than the former "hardcode a random number for each event" approach, this new approach is more brittle due to commonly used names. For example, multiple different developers could use *e.g.* `utils` as library name, and `hash` as event name. IIUC this would result in an event ID collision *e.g.* when trying to merge both mast forests.

This is mainly due to the fact that library names are not guaranteed to be unique.

**Comment by adr1anh** [comment]
> While cleaner than the former "hardcode a random number for each event" approach, this new approach is more brittle due to commonly used names. For example, multiple different developers could use _e.g._ `utils` as library name, and `hash` as event name. IIUC this would result in an event ID collision _e.g._ when trying to merge both mast forests.
> 
> This is mainly due to the fact that library names are not guaranteed to be unique.

I agree, and this is starting to touch on how events will work in the context of the network as a whole.

From what I understand, event handlers will be available as WASM binaries whose hash can be stored on-chain. The tooling and "package manage" can ensure that the user is aware of where these handlers originate from, and decide whether they want to trust them. We could even have a mode where the user can explicitly approve calling the handlers if they are sensitive (e.g. a signature handler).

One mechanism that could work well in this context would be to derive the ID from the hash of the WASM binary, allowing for non-colliding `"library_name::event_name"`. The disadvantage here would be that any updates to the handler would result in different event names. We could use the network though to store a static hash linking both a hash to derive a non-colliding event ID with the hash of the latest handler binary. 

I haven't fully thought through the details here, so there might be an easier way to ensure uniqueness. For now though, these are details that we can figure out later as long as we support any kind of string representation for IDs.

**Comment by bitwalker** [comment]
> I think it would be good to prepend the user-provided string with the namespace of the library in which the module is located. But other than that, I wouldn't do any transformations of the user-provided event name.

IMO we should leverage the existing namespacing rules, i.e. events are namespaced the same as procedures. We then add support for exporting/importing constants/events in addition to procedures, so everything works in a fairly uniform, predictable fashion. We'd add a new declaration type specifically for events, as I think it is a bad idea to try and force all of this into constant declarations, i.e. something like the following:

```
event.foo # defines the 'foo' event for local use
```

Exporting events would use the new syntax for controlling item visibility in MASM modules (i.e. move away from `export` and `proc` to `pub proc` and `proc`; and add support for `pub const` and when events are added, `pub event`).

It's likely exporting events would be relatively rare, but it sort of depends on what they are being used for IMO. We could disallow exporting events, but I think the choice of whether to export them or not should be left up to the owner of the event.

If someone wishes to, they can always emit an ostensibly internal event, simply by constructing the MAST themselves, or by pushing the event identifier on the stack by hand, so I don't think we should rely on visibility of events as some kind of security measure (I know that wasn't brought up, but I'm mentioning it now in case someone gets the idea).

> One mechanism that could work well in this context would be to derive the ID from the hash of the WASM binary, allowing for non-colliding "library_name::event_name". The disadvantage here would be that any updates to the handler would result in different event names. 

I think that's a significant problem unless the MAST which depends on the specific event handler implementation changes at the same time (and if all that is changing is the event handler, then the MAST is unlikely to change at all). IMO, either the event handlers and the MAST need to be bound together, or you need to use something less brittle than a hash.

**Comment by bobbinth** [comment]
> > I think it would be good to prepend the user-provided string with the namespace of the library in which the module is located. But other than that, I wouldn't do any transformations of the user-provided event name.
> 
> IMO we should leverage the existing namespacing rules, i.e. events are namespaced the same as procedures. We then add support for exporting/importing constants/events in addition to procedures, so everything works in a fairly uniform, predictable fashion. We'd add a new declaration type specifically for events, as I think it is a bad idea to try and force all of this into constant declarations, i.e. something like the following:
> 
> ```
> event.foo # defines the 'foo' event for local use
> ```
> 
> Exporting events would use the new syntax for controlling item visibility in MASM modules (i.e. move away from `export` and `proc` to `pub proc` and `proc`; and add support for `pub const` and when events are added, `pub event`).

Ah - that would be nice! Though it seems like a much bigger change and I'd rather do this later on. One of the motivations for the current event workstream is to unblock the work on precompiles (which we have external teams waiting for), but once this is done, we can come back and make events "first-class citizens".

Overall, I think something like this would be pretty cool:
```
# assuming we are in `my_lib::my_module` namespace, the event ID would be computed using
# `my_lib::my_module_my_event` string.

event.my_event

begin
    emit.my_event
end
```

One question that we'll need to answer is how would we get access to the ID of this event in supporting code. For example, our transaction host needs to know event IDs to properly handle custom events. Right now this is relatively easy because we hard-code event IDs both in MASM and in Rust. But once these IDs are automatically generated, we'll need a methodology to get event IDs at compile time from libraries somehow. But this too can be solved later on.

**Comment by bobbinth** [comment]
Closed by https://github.com/0xMiden/miden-vm/pull/2138 and other related PRs (we decided not to split event ID as was described in this issue).
